// @tensorflow/tfjs-backend-webgpu/dist/sparse_segment_reduce_webgpu.js
/**
 * @license
 * Copyright 2023 Google LLC.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * =============================================================================
 */
import { atomicAddSnippet } from './shader_util';
import { getMainHeaderString as main } from './webgpu_program';
import { computeDispatch, flatDispatchLayout } from './webgpu_util';
export class SparseSegmentSumProgram {
    constructor(outShape, sparseSize, outputDtype) {
        this.variableNames = ['input', 'indices', 'segmentIds'];
        this.outputShape = [];
        this.uniforms = 'segmentSize : i32, sparseSize : i32,';
        this.workgroupSize = [64, 1, 1];
        this.atomic = true;
        this.outputShape = outShape;
        this.type = outputDtype;
        this.dispatchLayout = flatDispatchLayout([sparseSize]);
        this.dispatch =
            computeDispatch(this.dispatchLayout, [sparseSize], this.workgroupSize);
        this.shaderKey = 'sparseSegmentSum';
    }
    getUserCode() {
        const userCode = `
    ${main('index')} {
      if (index < uniforms.sparseSize) {
        let indexInSegmentIds = index / uniforms.segmentSize;
        let indexInSegment = index % uniforms.segmentSize;
        let indexInInput = indices[indexInSegmentIds];
        let segmentId = segmentIds[indexInSegmentIds];

        let value = input[indexInInput * uniforms.segmentSize + indexInSegment];
        let outIndex = segmentId * uniforms.segmentSize + indexInSegment;
        ${atomicAddSnippet('&result[outIndex]', 'value', this.type)}
      }
    }
  `;
        return userCode;
    }
}
export class SparseSegmentIdCountProgram {
    constructor(outShape, segmentIdsShape) {
        this.variableNames = ['segmentIds'];
        this.outputShape = [];
        this.workgroupSize = [64, 1, 1];
        this.atomic = true;
        this.outputShape = [outShape];
        this.dispatchLayout = flatDispatchLayout(segmentIdsShape);
        this.dispatch = computeDispatch(this.dispatchLayout, segmentIdsShape, this.workgroupSize);
        this.shaderKey = 'sparseSegmentIdCountProgram';
    }
    getUserCode() {
        const userCode = `
    ${main('index')} {
      if (index < uniforms.segmentIdsShape) {
        let segmentId = segmentIds[index];
        ${atomicAddSnippet('&result[segmentId]', '1', 'int32')}
      }
    }
  `;
        return userCode;
    }
}
export class SparseSegmentMeanProgram {
    constructor(outShape, outputDtype) {
        this.variableNames = ['segmentSum', 'sameSegmentIdCount'];
        this.outputShape = [];
        this.uniforms = 'segmentSize : i32';
        this.workgroupSize = [64, 1, 1];
        this.size = true;
        this.outputShape = outShape;
        this.type = outputDtype;
        this.dispatchLayout = flatDispatchLayout(outShape);
        this.dispatch =
            computeDispatch(this.dispatchLayout, outShape, this.workgroupSize);
        this.shaderKey = 'sparseSegmentMean';
    }
    getUserCode() {
        const userCode = `
    ${main('index')} {
      if (index < uniforms.size) {
        let segmentId = index / uniforms.segmentSize;
        let count = sameSegmentIdCount[segmentId];
        if (count != 0) {
          ${this.type === 'float32' ?
            'setOutputAtIndex(index, segmentSum[index] / f32(count));' :
            'setOutputAtIndexI32(index, segmentSum[index] / count);'}
        }
      }
    }
  `;
        return userCode;
    }
}
